File: Task/K-d-tree/JavaScript/k-d-tree.js
/**
 * Class for representing a point. coordinate_type must be a numeric type.
 */
class Point {
  constructor(coords) {
    if (Array.isArray(coords)) {
      this.coords = [...coords];
    } else if (coords instanceof Object) {
      this.coords = Array.from(coords);
    }
  }

  /**
   * Returns the coordinate in the given dimension.
   *
   * @param {number} index dimension index (zero based)
   * @return {number} coordinate in the given dimension
   */
  get(index) {
    return this.coords[index];
  }

  /**
   * Returns the distance squared from this point to another point.
   *
   * @param {Point} pt another point
   * @return {number} distance squared from this point to the other point
   */
  distance(pt) {
    let dist = 0;
    for (let i = 0; i < this.coords.length; ++i) {
      const d = this.get(i) - pt.get(i);
      dist += d * d;
    }
    return dist;
  }

  toString() {
    return `(${this.coords.join(', ')})`;
  }
}

/**
 * JavaScript k-d tree implementation, based on the C++ version.
 */
class KDTree {
  constructor(pointsOrGenerator, count) {
    this.nodes = [];
    this.dimensions = 0;
    this.root = null;
    this.best = null;
    this.bestDist = 0;
    this.visited = 0;

    if (typeof pointsOrGenerator === 'function' && typeof count === 'number') {
      // Constructor with generator function
      for (let i = 0; i < count; ++i) {
        const point = pointsOrGenerator();
        if (i === 0) {
          this.dimensions = point.coords.length;
        }
        this.nodes.push({ point, left: null, right: null });
      }
    } else if (Array.isArray(pointsOrGenerator)) {
      // Constructor with array of points
      this.nodes = pointsOrGenerator.map(point => ({ point, left: null, right: null }));
      if (this.nodes.length > 0) {
        this.dimensions = this.nodes[0].point.coords.length;
      }
    }

    if (this.nodes.length > 0) {
      this.root = this.makeTree(0, this.nodes.length, 0);
    }
  }

  /**
   * Comparison function for sorting nodes by a specific dimension
   */
  nodeCmp(a, b, index) {
    return a.point.get(index) - b.point.get(index);
  }

  /**
   * Builds the tree recursively
   */
  makeTree(begin, end, index) {
    if (end <= begin) return null;

    const n = begin + Math.floor((end - begin) / 2);

    // Sort and find median
    this.nodes.slice(begin, end).sort((a, b) => this.nodeCmp(a, b, index));

    // Update index for the next level
    const nextIndex = (index + 1) % this.dimensions;

    // Recursively build left and right subtrees
    this.nodes[n].left = this.makeTree(begin, n, nextIndex);
    this.nodes[n].right = this.makeTree(n + 1, end, nextIndex);

    return this.nodes[n];
  }

  /**
   * Recursively finds the nearest node to the given point
   */
  findNearest(root, point, index) {
    if (!root) return;

    this.visited++;
    const d = root.point.distance(point);

    if (!this.best || d < this.bestDist) {
      this.bestDist = d;
      this.best = root;
    }

    if (this.bestDist === 0) return;

    const dx = root.point.get(index) - point.get(index);
    const nextIndex = (index + 1) % this.dimensions;

    // Search the side of the splitting plane that contains the point
    this.findNearest(dx > 0 ? root.left : root.right, point, nextIndex);

    // If the distance to the splitting plane is less than current best distance,
    // we need to check the other side too
    if (dx * dx >= this.bestDist) return;

    this.findNearest(dx > 0 ? root.right : root.left, point, nextIndex);
  }

  /**
   * Returns true if the tree is empty, false otherwise.
   */
  empty() {
    return this.nodes.length === 0;
  }

  /**
   * Returns the number of nodes visited by the last call to nearest().
   */
  getVisited() {
    return this.visited;
  }

  /**
   * Returns the distance between the input point and return value
   * from the last call to nearest().
   */
  getDistance() {
    return Math.sqrt(this.bestDist);
  }

  /**
   * Finds the nearest point in the tree to the given point.
   * It is not valid to call this function if the tree is empty.
   *
   * @param {Point} point a point
   * @return {Point} the nearest point in the tree to the given point
   */
  nearest(point) {
    if (!this.root) {
      throw new Error("Tree is empty");
    }

    this.best = null;
    this.visited = 0;
    this.bestDist = 0;

    this.findNearest(this.root, point, 0);

    return this.best.point;
  }
}

/**
 * Recreates the Wikipedia example from the original code
 */
function testWikipedia() {
  const points = [
    new Point([2, 3]),
    new Point([5, 4]),
    new Point([9, 6]),
    new Point([4, 7]),
    new Point([8, 1]),
    new Point([7, 2])
  ];

  const tree = new KDTree(points);
  const searchPoint = new Point([9, 2]);
  const nearest = tree.nearest(searchPoint);

  console.log("Wikipedia example data:");
  console.log("nearest point:", nearest.toString());
  console.log("distance:", tree.getDistance());
  console.log("nodes visited:", tree.getVisited());
}

/**
 * Generates random points and finds the nearest neighbor
 */
function randomPointGenerator(min, max) {
  return () => {
    const x = min + Math.random() * (max - min);
    const y = min + Math.random() * (max - min);
    const z = min + Math.random() * (max - min);
    return new Point([x, y, z]);
  };
}

function testRandom(count) {
  const rpg = randomPointGenerator(0, 1);
  const tree = new KDTree(rpg, count);
  const point = rpg();
  const nearest = tree.nearest(point);

  console.log(`Random data (${count} points):`);
  console.log("point:", point.toString());
  console.log("nearest point:", nearest.toString());
  console.log("distance:", tree.getDistance());
  console.log("nodes visited:", tree.getVisited());
}

// Main execution
try {
  testWikipedia();
  console.log();
  testRandom(1000);
  console.log();
  testRandom(10000); // Using 10,000 instead of 1,000,000 for browser performance
} catch (e) {
  console.error(e.message);
}

File: Task/K-d-tree/Dart/k-d-tree.dart
import 'dart:math';

// Define Point as a list of doubles
typedef Point = List<double>;

// Extension method for calculating squared distance
extension PointExtension on Point {
  double sqd(Point p) {
    if (length != p.length) {
      throw ArgumentError('Points must have the same dimension');
    }
    double sum = 0.0;
    for (int i = 0; i < length; i++) {
      double diff = this[i] - p[i];
      sum += diff * diff;
    }
    return sum;
  }
}

class HyperRect {
  final Point min;
  final Point max;

  HyperRect(this.min, this.max);

  // Create a copy with new lists for min and max
  HyperRect copy() => HyperRect(List.from(min), List.from(max));
}

class NearestNeighbor {
  final Point? nearest; // Nullable Point
  final double distSqd;
  final int nodesVisited;

  NearestNeighbor(this.nearest, this.distSqd, this.nodesVisited);
}

class KdNode {
  final Point domElt;
  final int split;
  KdNode? left; // Nullable and mutable
  KdNode? right; // Nullable and mutable

  KdNode(this.domElt, this.split, this.left, this.right);
}

class KdTree {
  final KdNode? n; // Root node (nullable)
  final HyperRect bounds;

  // Constructor builds the tree
  KdTree(List<Point> pts, this.bounds)
    : n = _buildTree(
        List.from(pts),
        0,
      ); // Pass a copy to avoid modifying original

  // Static helper method for recursive tree building (like nk2)
  static KdNode? _buildTree(List<Point> exset, int split) {
    if (exset.isEmpty) {
      return null;
    }

    // Sort the list in-place based on the current split dimension
    exset.sort((a, b) => a[split].compareTo(b[split]));

    int m = exset.length ~/ 2; // Integer division
    final Point d = exset[m];

    // Handle duplicate points at the median split value
    // Find the last index of the element equal to the median along the split dimension
    while (m + 1 < exset.length && exset[m + 1][split] == d[split]) {
      m++;
    }

    int s2 = split + 1;
    if (s2 == d.length) {
      s2 = 0; // Cycle through dimensions
    }

    // Dart's sublist creates copies, which is suitable here
    List<Point> leftSublist = exset.sublist(0, m);
    List<Point> rightSublist = exset.sublist(m + 1);

    return KdNode(
      d,
      split,
      _buildTree(leftSublist, s2),
      _buildTree(rightSublist, s2),
    );
  }

  // Public method to find the nearest neighbor
  NearestNeighbor nearest(Point p) {
    // Pass a copy of bounds because _nn might modify its HyperRect parameter
    return _nn(n, p, bounds.copy(), double.infinity);
  }

  // Private recursive helper for nearest neighbor search
  NearestNeighbor _nn(
    KdNode? kd, // Current node (nullable)
    Point target, // Target point
    HyperRect hr, // Hyper-rectangle for the current node's space
    double maxDistSqd, // Current best squared distance found so far
  ) {
    if (kd == null) {
      return NearestNeighbor(null, double.infinity, 0);
    }

    int nodesVisited = 1;
    final int s = kd.split;
    final Point pivot = kd.domElt;

    // Create copies for left and right potential subspaces
    final HyperRect leftHr = hr.copy();
    final HyperRect rightHr = hr.copy();

    // Adjust bounds for children
    leftHr.max[s] = pivot[s];
    rightHr.min[s] = pivot[s];

    // Determine which branch is nearer
    final bool targetInLeft = target[s] <= pivot[s];
    final KdNode? nearerKd = targetInLeft ? kd.left : kd.right;
    final HyperRect nearerHr = targetInLeft ? leftHr : rightHr;
    final KdNode? furtherKd = targetInLeft ? kd.right : kd.left;
    final HyperRect furtherHr = targetInLeft ? rightHr : leftHr;

    // Recurse down the nearer branch
    NearestNeighbor nearerResult = _nn(nearerKd, target, nearerHr, maxDistSqd);
    Point? nearest = nearerResult.nearest;
    double distSqd = nearerResult.distSqd;
    nodesVisited += nearerResult.nodesVisited;

    // Update the maximum distance squared if a closer point was found
    double maxDistSqd2 = (distSqd < maxDistSqd) ? distSqd : maxDistSqd;

    // Check the distance from the target to the splitting plane
    double d = pivot[s] - target[s];
    d *= d;

    // If the splitting plane is further than the current nearest, no need to check further branch
    if (d > maxDistSqd2) {
      return NearestNeighbor(nearest, distSqd, nodesVisited);
    }

    // Check distance from target to the pivot point itself
    d = pivot.sqd(target); // Use extension method
    if (d < distSqd) {
      nearest = pivot;
      distSqd = d;
      maxDistSqd2 = distSqd; // Update max distance again
    }

    // Recurse down the further branch if necessary
    NearestNeighbor furtherResult = _nn(
      furtherKd,
      target,
      furtherHr,
      maxDistSqd2,
    );
    nodesVisited += furtherResult.nodesVisited;

    // If the further branch found an even closer point
    if (furtherResult.distSqd < distSqd) {
      nearest = furtherResult.nearest;
      distSqd = furtherResult.distSqd;
    }

    return NearestNeighbor(nearest, distSqd, nodesVisited);
  }
}

// --- Helper Functions and Main ---

final rand = Random();

Point randomPt(int dim) => List.generate(dim, (_) => rand.nextDouble());

List<Point> randomPts(int dim, int n) => List.generate(n, (_) => randomPt(dim));

void showNearest(String heading, KdTree kd, Point p) {
  print('$heading:');
  print('Point            : $p'); // Default List.toString() is fine
  final result = kd.nearest(p);
  print('Nearest neighbor : ${result.nearest}');
  print(
    'Distance         : ${sqrt(result.distSqd)}',
  ); // Use sqrt from dart:math
  print('Nodes visited    : ${result.nodesVisited}');
  print(''); // Empty line
}

void main() {
  // Note: Using List<Point> instead of MutableList
  final points = <Point>[
    [2.0, 3.0],
    [5.0, 4.0],
    [9.0, 6.0],
    [4.0, 7.0],
    [8.0, 1.0],
    [7.0, 2.0],
  ];

  var hr = HyperRect([0.0, 0.0], [10.0, 10.0]);
  var kd = KdTree(points, hr); // Pass the original list
  showNearest('WP example data', kd, [9.0, 2.0]);

  hr = HyperRect([0.0, 0.0, 0.0], [1.0, 1.0, 1.0]);
  // Generate random points directly as List<Point>
  kd = KdTree(randomPts(3, 1000), hr);
  showNearest('1000 random 3D points', kd, randomPt(3));

  // Use hr.copy() if you intend to modify hr later, otherwise it's not strictly needed
  // The Kotlin code uses copy, so we replicate it here.
  var hrCopy = hr.copy();
  kd = KdTree(randomPts(3, 400000), hrCopy); // Pass the copy
  showNearest('400,000 random 3D points', kd, randomPt(3));
}
